### PR TITLE
Ebpf restoration

### PR DESCRIPTION
**Pull request by stp59**
changes:

- finish implementing the epbf architecture in `ebpf.ml`
- refactor entry point (CLI and STF testing) to select between different instantiations of the interpreter
- extend target interface with a `get_outport` function that grabs the egress port from the state after the pipeline is executed
- add newly supported ebpf test cases out of excluded. this is 10 new tests, one of which fails on a strange STF behavior
- proof of concept for the functorized interpreter!

### CODE REVIEW

**Comment by hackedy** [comment]
Haven't looked but the merge conflicts are probably from my recent change consolidating all the various parameter types into one type.

**Comment by stp59** [comment]
merged master in and everything is ok now

**Comment by hackedy** [comment]
Just making a note that `ebpf_checksum_extern.p4` and `ebpf_conntrack_extern.p4` are still excluded because they use the BMv2 custom C extern interface which we obviously will never support (#161).

**Comment by hackedy** [comment]
Added #230 for the 1 new failing test.

File: examples/checker_tests/good/hit_ebpf.stf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # bit<32> A bit<32> B
 # In the output B = (A + 10)
-add pipe_Check_src_ip 0 key.field0:0x0a019845 _NoAction()
-add pipe_Check_src_ip 0 key.field0:0x0a019846 pipe_Reject()
+add pipe.Check_src_ip 0 key.field0:0x0a019845 NoAction()
+add pipe.Check_src_ip 0 key.field0:0x0a019846 pipe.Reject(addr:0)
```
Comment (hackedy) [inline examples/checker_tests/good/hit_ebpf.stf:4]: Isn't this the same problem that issue870_ebpf.p4 was having?

File: lib/eval.ml [hunk 2]
```diff
@@ -364,77 +367,85 @@ module MakeInterpreter (T : Target) = struct
...
+                          ~f:(fun acc a -> if Types.name_eq (snd a).action.name action_name'
+                                           then (snd a).action.args
+                                           else acc)
+                          ~init:[] in
+      let ctrl_args = match_params_to_args type_params args |> List.map ~f:(convert_expression env) in
+      let pre_action_ref : Table.pre_action_ref =
+        { annotations = [];
+          name = action_name';
+          args = existing_args @ ctrl_args } in
+      let action : Table.typed_action_ref = { action = pre_action_ref; typ = Void } in (*type is a hack*)
```
Comment (hackedy) [inline lib/eval.ml:430]: We should have a validator/typechecker that turns `Ast` stuff into regular match/entry data using the typechecker environment but I think that might require some changes to the typechecker to make it easy to get at the types of actions that are nested inside a control etc.

File: lib/ebpf.ml [hunk 3]
```diff
@@ -94,7 +169,12 @@ module PreEbpfFilter : Target = struct
         eval_ebpf_ctrl ctrl filter [hdr_expr; accept_expr] app (env, st) in
       st, env, 
       if State.find_heap (EvalEnv.find_val accept_name env) st |> assert_bool
-      then Some (State.get_packet st) else None
+      then Some pkt
+      (* (State.get_packet st) *)
```
Comment (hackedy) [inline lib/ebpf.ml:173]: delete

File: stf-test/test.ml [hunk 4]
```diff
@@ -113,7 +151,23 @@ let stf_alco_test include_dir stf_file p4_file =
       let ic = In_channel.create stf_file in
       let lexbuf = Lexing.from_channel ic in
       let stmts = Test_parser.statements Test_lexer.token lexbuf in
-      run_test' include_dir p4_file stmts [] [] [] Eval.V1Interpreter.empty_state) in
+      let env, prog = 
+        Petr4_parse.parse_file include_dir p4_file false
+        |> (function `Ok p -> p | _ -> failwith "Petr4 parser error")
+        |> Elaborate.elab
+        |> Checker.check_program
```
Comment (hackedy) [inline stf-test/test.ml:158]: Doesn't typecheck. Maybe change to `|> fun (prog, renamer) -> Checker.check_program renamer prog`

File: lib/eval.ml (hunk 2 above)
Comment (stp59) [reply lib/eval.ml:430]: thats a good idea should we open a separate issue?

File: examples/checker_tests/good/hit_ebpf.stf (hunk 1 above)
Comment (stp59) [reply examples/checker_tests/good/hit_ebpf.stf:4]: yes; there were a lot of tests like that for some reason. i think p4c has a more flexible grammar for stf than we do, so i change a lot of these underscores to dots and added trivial control plane supplied arguments.

File: stf-test/test.ml (hunk 4 above)
Comment (stp59) [reply stf-test/test.ml:158]: it typechecks for me thats strange

Comment (hackedy) [reply stf-test/test.ml:158]: Ah, this was after merging master into it locally.  The elaborator returns a program and a renamer now

File: lib/eval.ml (hunk 2 above)
Comment (hackedy) [reply lib/eval.ml:430]: Yeah I'll do that right now

Comment (hackedy) [reply lib/eval.ml:430]: #232

**Pull request merged**